Allow INItializableWindow buttons/checkboxes to toggle visibility of other controls

Starkku opened the pull request:
Closes #775 

- It is now possible to set `$Toggles` on `XNAButton` or `XNACheckBox` & derivatives that are children of `INItializableWindow`. What this does is make the button/checkbox toggle visibility of the listed controls. Main use case would be additional panels for game options to conserve space where needed.

Starkku committed:
Add toggling control visibility option to INItializableWindow XNAClientButtons

FS-21 commented:
Since it has almost 1 year old I merged latest sources and this PR in a local repository.

For testing this variable $ToggleableControl that is mentioned in the main message I have declared both controls in;
```ini
[GameLobbyBase]
...
$CC29=SuperWeaponsOptionsPanel:XNAExtraPanel ;PlayerExtraOptionsPanel
$CC30=btnSuperWeaponsOptionsOpen:XNAClientButton
```
Then:
```ini
[btnSuperWeaponsOptionsOpen]
...
$ToggleableControl=SuperWeaponsOptionsPanel

[SuperWeaponsOptionsPanel]
$Width=400
$Height=200
$X=0
$Y=0; - LABEL_SPACING - DEFAULT_LBL_HEIGHT + 100
SolidColorBackgroundTexture=0,0,0,128
DrawMode=Stretched
BackgroundTexture=50.png
```

And works very well :-D

![new xna-client-feature-01](https://github.com/user-attachments/assets/3f3fad0f-504a-4e0b-a3ff-bcb348a7260a)

MahBoiDeveloper commented:
@Starkku merge https://github.com/Starkku/xna-cncnet-client/pull/9 to resolve merge conflict and continue development

Metadorius reviewed (changes requested):
Extra panel changes themselves are fine, some more questions on the rest.
Docs and migration are missing, you should fix that.

Metadorius commented on Docs/INISystem.md:
prob should be reverted and instead a migration note added

Starkku commented on DXMainClient/DXGUI/Multiplayer/PlayerExtraOptionsPanel.cs:
```diff
@@ -162,42 +162,42 @@ public override void Initialize()
 
             var btnClose = new XNAClientButton(WindowManager);
-            btnClose.Name = nameof(btnClose);
+            btnClose.Name = nameof(PlayerExtraOptionsPanel) + "_" + nameof(btnClose);
```
Prefix only or all of it? Client already uses `nameof` a lot for setting control `Name` which is used for INI parsing, but composite names are not really used so far.

Metadorius replied:
all of it I think, just so that refactors won't break the name. I agree it is counterintuitive but I feel in the long run it will be better

Starkku commented on ClientGUI/XNAClientButton.cs:
```diff
@@ -67,5 +67,35 @@ protected override void ParseControlINIAttribute(IniFile iniFile, string key, st
+
+            base.OnLeftClick();
+        }
+
```
I will be refactoring it so it is handled in `INItializableWindow.ReadLateAttributesForControl()` with event handlers and accepts multiple controls per button.

Starkku commented on ClientGUI/INItializableWindow.cs:
```diff
@@ -216,6 +216,10 @@ static string Localize(XNAControl control, string attributeName, string defaultV
+                else if (kvp.Key == "$ToggleableControl" && control is XNAClientButton button)
+                {
+                    button.SetToggleableControl(kvp.Value);
+                }
```
It could hypothetically be used to toggle visibility of game option controls e.g dropdowns/checkboxes directly, altho there may be some restrictions to this e.g only two view states per button (as it works as a binary toggle). As part of [further refactoring](https://github.com/CnCNet/xna-cncnet-client/pull/484#discussion_r2581542664) I can explicitly restrict it to `XNAExtraPanel`, `XNAPanel` and derivatives or leave the restrictions out entirely.

Syntax can be changed to your suggestions, it is even more appropriate if it is not restricted to just panels.

Starkku committed:
Refactor logic

- Remove bespoke logic from XNAClientButton
- Allow any control to be toggled
- Allow checkboxes to toggle controls
- Allow one control to toggle multiple controls

Starkku commented:
`PlayerExtraOptionsPanel` changes have been separated from this PR.

Other changes:
- Remove bespoke logic from XNAClientButton
- Allow any control to be toggled
- Allow checkboxes to toggle controls
- Allow one control to toggle multiple controls

Metadorius commented on ClientGUI/INItializableWindow.cs:
```diff
@@ -237,15 +239,52 @@ private void ReadLateAttributesForControl(XNAControl control)
+
+                    if (!string.IsNullOrWhiteSpace(toggles))
+                    {
+                        var controlnames = toggles.Split(',', StringSplitOptions.RemoveEmptyEntries);
```
IIRC we have an extension method for cleaned up list splitting

Metadorius commented on ClientGUI/UIHelpers.cs:
```diff
@@ -0,0 +1,57 @@
+    /// <param name="controlName">Name of the child control.</param>
+    /// <param name="recursive">Whether or not to look for children recursively.</param>
+    /// <returns>Child control matching the given name if found, otherwise type default value.</returns>
+    public static T FindMatchingChild<T>(XNAControl parent, string controlName, bool recursive = false)
```
perhaps better to use extensions and place this in an appropriately named file? both of those seem to be fitting as `XNAControl` extension methods

SadPencil replied:
e.g., in INItializableWindow we have defined FindChild and FindChildrenStartWith. Can follow that pattern and use an extension

SadPencil commented on ClientGUI/Settings/SettingCheckBoxBase.cs:
```diff
@@ -46,7 +44,7 @@ public string ParentCheckBoxName
             {
                 _parentCheckBoxName = value;
-                UpdateParentCheckBox(FindParentCheckBox());
+                UpdateParentCheckBox(UIHelpers.FindMatchingChild<XNAClientCheckBox>(Parent, _parentCheckBoxName, false));
```
I would leave FindParentCheckBox() method and write the find child statement into the FindParentCheckBox method body

MahBoiDeveloper committed:
Merge develop to the branch

MahBoiDeveloper committed:
Improve splitting

MahBoiDeveloper committed:
Refactor `UIHelpers` to the `XNAControlExtensions`

SadPencil commented on ClientGUI/Extensions/XNAControlExtensions.cs:
```diff
@@ -0,0 +1,57 @@
+    /// <param name="controlName">Name of the child control.</param>
+    /// <param name="recursive">Whether or not to look for children recursively.</param>
+    /// <returns>Child control matching the given name if found, otherwise type default value.</returns>
+    public static T FindMatchingChild<T>(this XNAControl parent, string controlName, bool recursive = false)
```
use name `FindChild`? it matches our existing method name

SadPencil committed:
Use the same FindChild implementation from INItializableWindow

SadPencil committed:
Remove the duplicate AnyChildMatches implementation

SadPencil reviewed (changes requested):
I don't think the name `$Toggles` is clear -- its name can suggest it toggles a checkbox to be checked or unchecked. Needs to think another name, probably `$VisibilityToggles`

SadPencil committed:
Parameter `recursive` should be turned off in UpdateParentCheckBox

SadPencil committed:
Rename `$Toggles`

SadPencil commented:
Renamed to `$VisibilityToggles` in a525ca7cd51cc13daef8191c90cf18f567e651b5. Feel free to suggest a better name

SadPencil committed:
Explain the optional parameter

SadPencil committed:
Remove unused using

SadPencil committed:
Avoid touching XNAClientButton.cs

SadPencil commented on ClientGUI/INItializableWindow.cs:
```diff
@@ -266,15 +221,52 @@ private void ReadLateAttributesForControl(XNAControl control)
+                                    {
+                                        toggleControl.Enabled = !toggleControl.Enabled;
+                                        toggleControl.Visible = !toggleControl.Visible;
+                                    };
```
```cs
                                    child.LeftClick += (sender, args) =>
                                    {
                                        toggleControl.Enabled = !toggleControl.Enabled;
                                        toggleControl.Visible = !toggleControl.Visible;
                                    };
```

Uhh... in this case the toggleControl's visibility is *binded* to the button. `child.LeftClick` does not observe `toggleControl.Enabled/Visible` changes so possible wrong event handler was listened on.

Same for `checkBox.CheckedChanged`.

Therefore, I think we can't merge this PR right now. We need to carefully think of its usage, and provide an example. Also, the name should probably be `BindsVisibilityTo` or `BindsVisibilityFrom` depends on the implementation.

SadPencil committed:
Merge branch 'develop' into pr/484

SadPencil commented:
> Uhh... in this case the toggleControl's visibility is binded to the button. child.LeftClick does not observe toggleControl.Enabled/Visible changes so possible wrong event handler was listened on.
>
> Same for checkBox.CheckedChanged.
>
> Therefore, I think we can't merge this PR right now. We need to carefully think of its usage, and provide an example. Also, the name should probably be BindsVisibilityTo or BindsVisibilityFrom depends on the implementation.

For example, maybe we can change to bind, whether a checkbox is *checked* (or *unchecked*), the visibility of the target controls are changed. Like `$BindsVisibilityTo=controlName1, controlName2, !controlName3`

More discussions are welcomed.

MahBoiDeveloper committed:
Reduce nesting
